events/controller.go

```diff
@@ -0,0 +1,76 @@
+package events
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	pmetrics "github.com/awslabs/operatorpkg/metrics"
+	"github.com/awslabs/operatorpkg/object"
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/clock"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type Controller[T client.Object] struct {
+	kind       string
+	startTime  time.Time
+	kubeClient client.Client
+	eventCount pmetrics.CounterMetric
+}
+
+func NewController[T client.Object](client client.Client, clock clock.Clock) *Controller[T] {
+	kind := object.GVK(object.New[T]()).Kind
+	EventCountOpts.Subsystem = strings.ToLower(kind)
+	return &Controller[T]{
+		kind:       kind,
+		startTime:  clock.Now(),
+		kubeClient: client,
+		eventCount: pmetrics.NewPrometheusCounter(
+			metrics.Registry,
+			EventCountOpts,
+			EventCountLabels,
+		),
+	}
+}
+
+func (c *Controller[T]) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		For(&v1.Event{}, builder.WithPredicates(predicate.NewTypedPredicateFuncs(func(o client.Object) bool {
+			// Only reconcile on the object kind we care about
+			event := o.(*v1.Event)
+			return event.InvolvedObject.Kind == c.kind
```

Comment:
Should we also check the group? What happens if there are two resources on the cluster that have the same kind and different groups?